maggma/examples/builders.py

```diff
@@ -0,0 +1,274 @@
+
+        Args:
+            item: item to process
+
+        Returns:
+            dict: a dict that will extend a dict pre-populated with
+                {self.source.key, self.course.lu_field} fields.
+        """
+
+    def get_items(self):
+        criteria = get_criteria(
+            self.source, self.target, query=self.query,
+            incremental=self.incremental, logger=self.logger)
+        return self.source.query(criteria=criteria)
+
+    def process_item(self, item):
+        try:
+            processed = self.ufn.__call__(item)
+        except Exception as e:
+            self.logger.error(traceback.format_exc())
+            processed = {"error": str(e)}
+        key, lu_field = self.source.key, self.source.lu_field
+        out = {key: item[key], lu_field: item[lu_field]}
+        out.update(processed)
+        return out
+
+    def update_targets(self, items):
+        source, target = self.source, self.target
+        for item in items:
+            # Use source last-updated value, ensuring `datetime` type.
+            item[target.lu_field] = source.lu_func[0](item[source.lu_field])
+            if source.lu_field != target.lu_field:
+                del item[source.lu_field]
+            item["_bt"] = datetime.utcnow()
+            if "_id" in item:
+                del item["_id"]
+        target.update(items, update_lu=False)
+
+
+class GroupBuilder(MapBuilder, metaclass=ABCMeta):
+    """
+    Group source docs and produce one target doc from each group.
+
+    Supports incremental building, where a source group gets (re)built only if
+    it has a newer (by lu_field) doc than the corresponding (by key) target doc.
+    """
+    def __init__(self, source, target, query=None, incremental=True, **kwargs):
+        """
+
+        Given criteria, get docs with needed grouping properties. With these
+        minimal docs, yield groups. For each group, fetch all needed data for
+        item processing, and yield one or more items (i.e. subgroups as
+        appropriate).
+
+        Args:
+            source (Store): source store
+            target (Store): target store
+            query (dict): optional query to filter source store
+            incremental (bool): whether to use lu_field of source and target
+                to get only new/updated documents.
+        """
+        super().__init__(
+            source, target, query=query, incremental=incremental, **kwargs)
+        self.total = None
+
+    def get_items(self):
+        criteria = get_criteria(
+            self.source, self.target, query=self.query,
+            incremental=self.incremental, logger=self.logger)
+        if all(isinstance(entry, str) for entry in self.grouping_properties()):
+            properties = {entry: 1 for entry in self.grouping_properties()}
+            if "_id" not in properties:
+                properties.update({"_id": 0})
+        else:
+            properties = {entry: include for entry, include in self.projection}
```

Comment:
Where does self.projection come from?

Reply:
oops, a holdover from when I had the class attribute `projection` versus the method `grouping_properties`. Will fix.

Reply:
Done with 89f7c1f